hatasmota/camera.py

```diff
@@ -0,0 +1,141 @@
+            mac=config[CONF_MAC],
+            platform=platform,
+            poll_payload="",
+            poll_topic=get_topic_command_state(config),
+            availability_topic=get_topic_tele_will(config),
+            availability_offline=config_get_state_offline(config),
+            availability_online=config_get_state_online(config),
+            deep_sleep_enabled=config[CONF_DEEP_SLEEP],
+            command_topic=get_topic_command(config),
+            result_topic=get_topic_stat_result(config),
+            state_topic=get_topic_tele_state(config),
+            sensor_topic=get_topic_tele_sensor(config),
+            ip_address=config[CONF_IP]
+        )
+
+
+class TasmotaCamera(TasmotaAvailability, TasmotaEntity):
+    """Representation of a Tasmota camera."""
+
+    _cfg: TasmotaCameraConfig
+
+    def __init__(self, **kwds: Any):
+        """Initialize."""
+        self._sub_state: dict | None = None
+        super().__init__(**kwds)
+
+    async def subscribe_topics(self) -> None:
+        """Subscribe to topics."""
+
+        def state_message_received(msg: ReceiveMessage) -> None:
+            """Handle new MQTT state messages."""
+            if not self._on_state_callback:
+                return
+
+            if msg.topic == self._cfg.sensor_topic:
+                state = get_value_by_path(msg.payload, ["CAMERA"])
+                if state:
+                    self._on_state_callback(state)
+
+
+        availability_topics = self.get_availability_topics()
+        topics = {
+            "result_topic": {
+                "event_loop_safe": True,
+                "topic": self._cfg.result_topic,
+                "msg_callback": state_message_received,
+            },
+            "state_topic": {
+                "event_loop_safe": True,
+                "topic": self._cfg.state_topic,
+                "msg_callback": state_message_received,
+            },
+            "sensor_topic": {
+                "event_loop_safe": True,
+                "topic": self._cfg.sensor_topic,
+                "msg_callback": state_message_received,
+            },
+        }
+        topics = {**topics, **availability_topics}
+
+        self._sub_state = await self._mqtt_client.subscribe(
+            self._sub_state,
+            topics,
+        )
+
+    async def unsubscribe_topics(self) -> None:
+        """Unsubscribe to all MQTT topics."""
+        self._sub_state = await self._mqtt_client.unsubscribe(self._sub_state)
+
+    def get_still_image_stream(self, websession:aiohttp.ClientSession) -> web.StreamResponse | None:
+        """Get the io stream to read the static image."""
+        still_image_url = f"http://{self._cfg.ip_address}/snapshot.jpg"
+        return websession.get(still_image_url)
+
+    def get_mjpeg_stream(self, websession:aiohttp.ClientSession) -> web.StreamResponse | None:
+        """Get the io stream to read the mjpeg stream."""
+        mjpeg_url = f"http://{self._cfg.ip_address}:81/cam.mjpeg"
```

Comment:
Same question, is the port always 81 and the url always `cam.mjpeg` for the mjpeg stream?

Reply:
From the [Tasmota code](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam.ino#L1110) (or [new version](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino#189)),
The port is fixed to 81.

we have a choice of `/cam.mjpeg`, `/cam.jpg`, `/stream`. IMO, `stream` and `cam.mjpeg` are good choices among these. I chose `cam.mjpeg`, because it clarifies that the stream is an `mjpeg` stream.

Reply:
Same question as above; all the URLs just call the same function `HandleWebcamMjpeg`, does it mean there's no difference, it's just aliases for exactly the same funcitonality?

Reply:
There is no difference. All these URLs have the same callback function and return the same data.
I guess the Tasmota authors added them only for convenience.